crates/rust-analyzer/src/handlers/notification.rs

```diff
@@ -287,16 +288,33 @@ fn run_flycheck(state: &mut GlobalState, vfs_path: VfsPath) -> bool {
         let world = state.snapshot();
         let mut updated = false;
         let task = move || -> std::result::Result<(), ide::Cancelled> {
-            // Trigger flychecks for all workspaces that depend on the saved file
-            // Crates containing or depending on the saved file
-            let crate_ids: Vec<_> = world
-                .analysis
-                .crates_for(file_id)?
-                .into_iter()
-                .flat_map(|id| world.analysis.transitive_rev_deps(id))
-                .flatten()
-                .unique()
-                .collect();
+            // Is the target binary? If so we let flycheck run only for the workspace that contains the crate.
+            let target_is_bin = TargetSpec::for_file(&world, file_id)?.and_then(|x| {
+                if x.target_kind() == project_model::TargetKind::Bin {
+                    return match x {
+                        TargetSpec::Cargo(c) => Some(c.target),
+                        TargetSpec::ProjectJson(p) => Some(p.label),
+                    };
+                }
+
+                None
+            });
+
+            let crate_ids = if target_is_bin.is_some() {
+                // Trigger flychecks for the only workspace which the binary crate belongs to
+                world.analysis.crates_for(file_id)?.into_iter().unique().collect::<Vec<_>>()
```

Comment:
We should already have that information in the `TargetSpec`, it should contain the relevant crate id

Reply:
this is the only bit I couldn't really come up with a solution because the package information for `RustProjectJson` is not really clear.

Reply:
No worries, https://github.com/rust-lang/rust-analyzer/pull/17946 the data was there but not exposed that's all